### PR TITLE
Fix creating datasets from nested lists of str/bytes objects & writing to fixed UTF-8 datasets

### PR DESCRIPTION
**Pull request by takluyver**
Fix creating datasets from nested lists of str/bytes objects.

These now behave the same as using a single object or a simple list. See https://github.com/h5py/h5py/issues/1338#issuecomment-667538812 .

EDIT: Now also allows storing str objects with any unicode characters into a fixed-length UTF-8 string dataset. Previously, str objects with non-ASCII characters would error.

### CODE REVIEW

File: h5py/_hl/base.py [hunk 1]
```diff
@@ -75,10 +81,7 @@ def guess_dtype(data):
         if isinstance(data, h5r.Reference):
             return h5t.ref_dtype
 
-        item_type = object_collection_of(data)
-        if item_type is None:
-            # Potentially scalar
-            item_type = type(data)
+        item_type = find_item_type(data)
```
Comment (tacaswell) [inline h5py/_hl/base.py:84]: There is a path through `find_item_type` that returns `None`, but the code that used to be here ensured that `item_type` was never `None` before going into the rest of the code.

Comment (takluyver) [reply h5py/_hl/base.py:84]: It did, but then the only things we do with item_type are to check `item_type is bytes` and `item_type is str` immediately below. The case which the extra `if item_type is None` previously handled - a single str/bytes object - is now handled inside `find_item_type()`.

`item_type` will be None if there is a non-homogeneous collection (different types in a list) or a numpy array with a more specific dtype already.

**Comment by aragilar** [comment]
This looks good to me.

**Comment by takluyver** [comment]
Thanks! I'll wait until the US maintainers can take a look later today, but I'll merge it if there are no objections. I think this completes the string work for 3.0. :slightly_smiling_face:

**Pull request merged**